### PR TITLE
mangohud: init at 0.4.1

### LINKED ISSUE #231: Build failure on Musl: type of 'readlink' is unknown
I'm trying to build MangoHud for Alpine Linux which uses Musl rather than glibc, and it seems it fails to build because of it.

```
ninja: job failed: g++ -Isrc/25a6634@@MangoHud@sha -Isrc -I../src -I../include -I../modules/ImGui/src -I. -I/usr/include/dbus-1.0 -I/usr/lib/dbus-1.0/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++14 -Werror=return-type -fno-math-errno -fno-trapping-math -Wno-non-virtual-dtor -Wno-missing-field-initializers -Wno-format-truncation -Os -fomit-frame-pointer -Os -fomit-frame-pointer -fPIC -pthread -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS '-DPACKAGE_VERSION="v0.4.0"' -DNDEBUG -D_GNU_SOURCE -DHAVE_PTHREAD -DUSE_GCC_ATOMIC_BUILTINS -DHAVE_TIMESPEC_GET -DHAVE___BUILTIN_BSWAP32 -DHAVE___BUILTIN_BSWAP64 -DHAVE___BUILTIN_CLZ -DHAVE___BUILTIN_CLZLL -DHAVE___BUILTIN_CTZ -DHAVE___BUILTIN_EXPECT -DHAVE___BUILTIN_FFS -DHAVE___BUILTIN_FFSLL -DHAVE___BUILTIN_POPCOUNT -DHAVE___BUILTIN_POPCOUNTLL -DHAVE___BUILTIN_UNREACHABLE '-DMANGOHUD_ARCH="64bit"' -DHAVE_NVML -DHAVE_X11 -DHAVE_DBUS -fvisibility=hidden -DVK_USE_PLATFORM_XLIB_KHR -MD -MQ 'src/25a6634@@MangoHud@sha/file_utils.cpp.o' -MF 'src/25a6634@@MangoHud@sha/file_utils.cpp.o.d' -o 'src/25a6634@@MangoHud@sha/file_utils.cpp.o' -c ../src/file_utils.cpp
In file included from /usr/include/fortify/strings.h:22,
                 from /usr/include/string.h:57,
                 from /usr/include/fortify/wchar.h:30,
                 from /usr/include/c++/9.3.0/cwchar:44,
                 from /usr/include/c++/9.3.0/bits/postypes.h:40,
                 from /usr/include/c++/9.3.0/bits/char_traits.h:40,
                 from /usr/include/c++/9.3.0/string:40,
                 from ../src/file_utils.h:2,
                 from ../src/file_utils.cpp:1:
/usr/include/fortify/unistd.h:119:1: error: type of 'readlink' is unknown
  119 | _FORTIFY_FN(readlink) ssize_t readlink(const char *__p, char *__s, size_t __n)
      | ^~~~~~~~~~~
In file included from ../src/file_utils.cpp:5:
/usr/include/fortify/unistd.h: In function 'ssize_t readlink(const char*, char*, size_t)':
/usr/include/fortify/unistd.h:125:38: error: '__orig_readlink' cannot be used as a function
  125 |  return __orig_readlink(__p, __s, __n);
      |                                      ^
../src/overlay.cpp: In function 'void render_benchmark(swapchain_stats&, overlay_params&, ImVec2&, unsigned int, uint64_t)':
../src/overlay.cpp:1047:87: warning: format not a string literal and no format arguments [-Wformat-security]
 1047 |    ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, alpha / params.background_alpha), finished);
      |                                                                                       ^
../src/overlay.cpp:1052:79: warning: format not a string literal and no format arguments [-Wformat-security]
 1052 |    ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, alpha / params.background_alpha), duration);
      |                                                                               ^~~~~~~~
ninja: subcommand failed
```

### LINKED ISSUE #84686: MangoHud
**Project description**

A modification of the Mesa Vulkan overlay. Including GUI improvements, temperature reporting, and logging capabilities.

**Metadata**

* homepage URL: https://github.com/flightlessmango
* source URL: https://github.com/flightlessmango/MangoHud
* license: gpl3
* platforms: linux

### PR DESCRIPTION
**Pull request by zeratax**
###### Motivation for this change

closes #84686 

###### Things done

I tested this with several OpenGL and Vulkan 32/64 bit games.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Questions

This is currently a little hackish with the `combined.nix` file, since to use mangohud with 32bit application like a lot of games still are we also need to compile a 32bit version.
but if I just try to build the `default.nix` file with:
```nix
nix-env -f . -iA mangohud
nix-env -f . -iA mangohud --argstr system i686-linux
```

The second derivation fails with:
```
error: packages '/nix/store/98anspysbisprqmzr2rwsm9dvgkxzz2z-mangohud-0.4.1/share/vulkan/registry/cgenerator.py' and '/nix/store/hg7vqhxy0kb2kdxm6rrncajb036mjg15-vulkan-headers-1.2.131.1/share/vulkan/registry/cgenerator.py' have the same priority 5; use 'nix-env --set-flag priority NUMBER INSTALLED_PKGNAME' to change the priority of one of the conflicting packages (0 being the highest priority)
```

The `combined.nix` now installs both versions next to each other.

~~I am also not using the exact 0.4.1 version, but am a few commits ahead to fix https://github.com/flightlessmango/MangoHud/issues/231 and also https://github.com/flightlessmango/MangoHud/commit/acf2d88fbcb7a7a47f957a87d20739c67049bd0d~~

~~Or should I include all extra commits with the patches option?~~

Edit: resolved with https://github.com/ZerataX/nixpkgs/pull/1

### CODE REVIEW

File: pkgs/tools/graphics/mangohud/default.nix
Comment (MetaDark) [inline pkgs/tools/graphics/mangohud/default.nix:34]: Was there a problem building with xnvctrl?

**Review by kira-bruneau — COMMENTED** [review]
Thanks for doing this @ZerataX!

File: pkgs/tools/graphics/mangohud/default.nix
Comment (MetaDark) [inline pkgs/tools/graphics/mangohud/default.nix:68]: Description shouldn't have a period at the end: https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md

Comment (MetaDark) [inline pkgs/tools/graphics/mangohud/default.nix:68]: Description should not have a period at the end: https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md

**Comment by MetaDark** [comment]
There may be a better approach that I don't know about for combining the two builds, but this seems to work. I've also seen other projects use a later rev to get the build working, so I think it should be fine here.

I noticed that you had `Tested using sandboxing` and `Tested execution of all binary files` unchecked. Is this correct? It sounds like you may have already completed those.

**Comment by ZerataX** [comment]
@MetaDark right yes that was incorrect and I now checked those boxes. I also have a nur package here that seems to build fine: https://travis-ci.com/github/ZerataX/nur-packages
https://github.com/ZerataX/nur-packages/blob/master/default.nix#L19

File: pkgs/tools/graphics/mangohud/default.nix
Comment (ZerataX) [reply pkgs/tools/graphics/mangohud/default.nix:34]: I got it building now with this option enabled. I already included it in my nur package and will update it here later today:
https://github.com/ZerataX/nur-packages/blob/master/pkgs/MangoHud/default.nix#L34

Comment (ZerataX) [reply pkgs/tools/graphics/mangohud/default.nix]: should now be resolved, though I don't have the hardware to test if this works with nvidia gpus.

**Review by Atemu — COMMENTED** [review]
Result of `nixpkgs-review pr 95225` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
<br>- mangohud
</details>

When I try to run something with it I get a bunch of

```
MANGOHUD: Failed to open 64bit libnvidia-ml.so.1: libnvidia-ml.so.1: cannot open shared object file: No such file or directory
MANGOHUD: Failed to open 64bit libXNVCtrl.so.0: libXNVCtrl.so.0: cannot open shared object file: No such file or directory
```

though and it doesn't work :/

**Comment by ZerataX** [comment]
@Atemu so I imagine you are using an nvidia gpu? also would love to know what you tried to run with it? easiest to test is imo `vkcube` which comes with `vulkan-tools`. 32bit applications in this version of mangohud require to be run with `mangohud.x86`

**Comment by Atemu** [comment]
Yes it's an nvidia GPU, sorry should've mentioned that.

I tried Guild Wars 2 but these messages even appear when I try to run bash. Same story with vkcube.

**Review by Atemu — APPROVED** [review]
Aha! Got it to work. A nix-shell isn't enough, mangohud needs to be in your user env.

The missing library messages still show up weirdly but that's a cosmetic issue we can fix later.

File: pkgs/tools/graphics/mangohud/default.nix
Comment (ZerataX) [reply pkgs/tools/graphics/mangohud/default.nix]: ah yeah thanks!

**Review by Atemu — CHANGES_REQUESTED** [review]
https://github.com/ZerataX/nixpkgs/pull/1

**Review by Atemu — APPROVED** [review]

**Review by kira-bruneau — APPROVED** [review]

File: pkgs/tools/graphics/mangohud/combined.nix
Comment (samuelgrf) [inline pkgs/tools/graphics/mangohud/combined.nix:6]:
```suggestion
  mangohud_64 = pkgs.callPackage ./default.nix { inherit libXNVCtrl; };
  mangohud_32 = pkgs.pkgsi686Linux.callPackage ./default.nix { inherit libXNVCtrl; };
```

**Review by samuelgrf — APPROVED** [review]

**Comment by ZerataX** [comment]
I was wondering what the current status of this is? Do I just need to rebase this so the commit looks properly (not quite sure how I would go about doing that honestly) or do we wait for #101597 and go for a current release of mangohud?

**Comment by Atemu** [comment]
You should rebase on top of master or unstable and then squash the remaining two commits.

I'd be in favour of merging this as-is and not trying to perfect it by packaging the newest version.  
Fixing the bug in our ld.so that prevents the upgrade is _way_ outside this PR's scope I think and will probably require a core maintainer to have a look at. Given how busy they are with more important things than an obscure feature of a core library which 2/70000 packages would need, I wouldn't count on that happening anytime soon.  
Same goes for @MetaDark's #92401.

We could try would be to patch the new Mangohud to use the v4 loader again that we did manage to get working here but I think that should happen in a separate PR.

File: pkgs/tools/graphics/mangohud/combined.nix
Comment (Atemu) [inline pkgs/tools/graphics/mangohud/combined.nix:17]: Could you put these onto a single or two lines? It looks a bit ugly (especially the lone `;`) and we're not even close to 80chars.

Comment (ZerataX) [reply pkgs/tools/graphics/mangohud/combined.nix]: okay I think I manged to rebase without doing something horrible!

**Review by Atemu — APPROVED** [review]

File: pkgs/tools/graphics/mangohud/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/graphics/mangohud/default.nix:61]:
```suggestion
    pkg-config
```

**Comment by legendofmiracles** [comment]
Will you pick this up again @ZerataX?

**Comment by ZerataX** [comment]
hmm not sure how to squash that merge and my commit properly

File: pkgs/tools/graphics/mangohud/combined.nix
Comment (SuperSandro2000) [inline pkgs/tools/graphics/mangohud/combined.nix:9]:
```suggestion
  pname = "mangohud";
  version = mangohud_64.version;
```

File: pkgs/tools/graphics/mangohud/combined.nix
Comment (SuperSandro2000) [inline pkgs/tools/graphics/mangohud/combined.nix:11]:
```suggestion
  paths = [ mangohud_32 ] ++ lib.optionals stdenv.is64bit [ mangohud_64 ];
```

File: pkgs/tools/graphics/mangohud/combined.nix
Comment (ZerataX) [reply pkgs/tools/graphics/mangohud/combined.nix:9]: I don't think that's possible in `buildEnv`?

File: pkgs/tools/graphics/mangohud/default.nix
Comment (legendofmiracles) [inline pkgs/tools/graphics/mangohud/default.nix:22]: newest version is 0.6.1

**Comment by legendofmiracles** [comment]
> Edit: wow I did it!
Nice!
Feel free to add me as a maintainer

**Comment by legendofmiracles** [comment]
Result of `nixpkgs-review pr 95225` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>mangohud</li>
  </ul>
</details>

**Comment by legendofmiracles** [comment]
Well, this is awkward, running the `nix-review` command again worked flawlessly

**Comment by legendofmiracles** [comment]
Result of `nixpkgs-review pr 95225` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mangohud</li>
  </ul>
</details>

File: pkgs/tools/graphics/mangohud/combined.nix
Comment (legendofmiracles) [reply pkgs/tools/graphics/mangohud/combined.nix:9]: If you rename it to pname it prints this:
```
error: anonymous function at /home/nix/nixpkgs/pkgs/build-support/buildenv/default.nix:8:2 called without required argument 'name'

       at /home/nix/nixpkgs/lib/customisation.nix:69:16:

           68|     let
           69|       result = f origArgs;
             |                ^
           70|
```

File: pkgs/tools/graphics/mangohud/default.nix
Comment (Atemu) [reply pkgs/tools/graphics/mangohud/default.nix:22]: Versions newer than v4 won't work in Nixpkgs, see https://github.com/NixOS/nixpkgs/pull/95225#issuecomment-734643180

File: pkgs/tools/graphics/mangohud/combined.nix
Comment (SuperSandro2000) [reply pkgs/tools/graphics/mangohud/combined.nix:9]: > I don't think that's possible in buildEnv?

I did not know that.

**Comment by davidak** [comment]
Result of `nixpkgs-review pr 95225` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mangohud</li>
  </ul>
</details>

**Comment by davidak** [comment]
Tested using glxgears from `mesa-demos`.

```
[nix-shell:~/.cache/nixpkgs-review/pr-95225]$ mangohud glxgears
ERROR: ld.so: object 'libMangoHud.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS64): ignored.
MANGOHUD: Failed to open 64bit libnvidia-ml.so.1: libnvidia-ml.so.1: cannot open shared object file: No such file or directory
MANGOHUD: Failed to open 64bit libXNVCtrl.so.0: libXNVCtrl.so.0: cannot open shared object file: No such file or directory
MANGOHUD: Failed to open 64bit libX11.so.6: libX11.so.6: cannot open shared object file: No such file or directory
MANGOHUD: Failed to open 64bit libnvidia-ml.so.1: libnvidia-ml.so.1: cannot open shared object file: No such file or directory
MANGOHUD: Failed to open 64bit libXNVCtrl.so.0: libXNVCtrl.so.0: cannot open shared object file: No such file or directory
skipping config: /nix/store/5bx1vwh2v9dxbz5dss3mllrx6q5b7l0n-mesa-demos-8.4.0/bin/MangoHud.conf [ not found ]
skipping config: /home/davidak/.config/MangoHud/glxgears.conf [ not found ]
skipping config: /home/davidak/.config/MangoHud/MangoHud.conf [ not found ]
MANGOHUD: can't find file pci.ids
Version: 4.6 
Running synchronized to the vertical refresh.  The framerate should be
approximately the same as the monitor refresh rate.
329 frames in 5.0 seconds = 65.647 FPS
300 frames in 5.0 seconds = 59.890 FPS
...
```

Screenshot from 2021-04-26 18-43-27

Then installed it and tested in Steam by adding `mangohud %command%` to CS:GO launch options.

`[nix-shell:~/.cache/nixpkgs-review/pr-95225]$ nix-env -f nixpkgs/ -i mangohud`

That does also work.

**Review by davidak — APPROVED** [review]

**Comment by MetaDark** [comment]
It looks like the latest version is now 0.6.1, so it would probably be best to update it here.

**Comment by ZerataX** [comment]
> 
> 
> It looks like the latest version is now [0.6.1](https://github.com/flightlessmango/MangoHud/releases/tag/v0.6.1), so it would probably be best to update it here.

yeah we decided to do this in a different pr https://github.com/NixOS/nixpkgs/pull/95225#issuecomment-734643180

**Review by kira-bruneau — APPROVED** [review]
> yeah we decided to do this in a different pr [#95225 (comment)](https://github.com/NixOS/nixpkgs/pull/95225#issuecomment-734643180)

Oh right thanks, I forgot about that :sweat_smile:. This looks good to me then, thanks @ZerataX!

**Pull request merged**
